File: android/src/main/java/expo/modules/liveactivity/ExpoLiveActivityModule.kt
package expo.modules.liveactivity

import android.content.Context
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.liveactivity.core.services.LiveActivityServiceProtocol
import expo.modules.liveactivity.core.services.LiveActivityService
import expo.modules.liveactivity.core.models.*
import expo.modules.liveactivity.core.usecases.*
import expo.modules.liveactivity.infrastructure.repositories.InMemoryActivityRepository
import expo.modules.liveactivity.infrastructure.repositories.ActivityRepositoryProtocol
import expo.modules.liveactivity.infrastructure.notifications.NotificationActivityManager
import expo.modules.liveactivity.infrastructure.notifications.NotificationActivityManagerProtocol
import expo.modules.liveactivity.infrastructure.services.AndroidNotificationService
import expo.modules.liveactivity.infrastructure.services.AndroidNotificationServiceProtocol
import expo.modules.liveactivity.presentation.events.ActivityEventPublisher
import expo.modules.liveactivity.presentation.events.ActivityEventPublisherProtocol
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.*

class ExpoLiveActivityModule : Module() {
    
    // MARK: - Dependencies (Lazy Initialization)
    
    private val androidContext: Context by lazy {
        this.appContext.reactContext ?: this.appContext.currentActivity ?: throw IllegalStateException("App context is null")
    }
    
    private val notificationManager: NotificationActivityManagerProtocol by lazy {
        NotificationActivityManager(androidContext)
    }
    
    private val repository: ActivityRepositoryProtocol by lazy {
        InMemoryActivityRepository()
    }
    
    private val eventPublisher: ActivityEventPublisherProtocol by lazy {
        ActivityEventPublisher()
    }
    
    private val validator: ActivityConfigValidatorProtocol by lazy {
        ActivityConfigValidator()
    }
    
    private val notificationService: AndroidNotificationServiceProtocol by lazy {
        AndroidNotificationService(androidContext, notificationManager)
    }
    
    private val liveActivityService: LiveActivityServiceProtocol by lazy {
        LiveActivityService(
            notificationManager = notificationManager,
            repository = repository,
            eventPublisher = eventPublisher
        )
    }
    
    // Use Cases
    private val startActivityUseCase: StartActivityUseCaseProtocol by lazy {
        StartActivityUseCase(
            activityService = liveActivityService,
            repository = repository,
            validator = validator,
            eventPublisher = eventPublisher
        )
    }
    
    private val updateActivityUseCase: UpdateActivityUseCaseProtocol by lazy {
        UpdateActivityUseCase(
            activityService = liveActivityService,
            repository = repository,
            eventPublisher = eventPublisher
        )
    }
    
    private val endActivityUseCase: EndActivityUseCaseProtocol by lazy {
        EndActivityUseCase(
            activityService = liveActivityService,
            repository = repository,
            eventPublisher = eventPublisher
        )
    }
    
    private val coroutineScope = CoroutineScope(Dispatchers.Main)
    
    // MARK: - Module Definition
    
    override fun definition() = ModuleDefinition {
        Name("ExpoLiveActivity")
        
        // Constants
        Constants(
            "isSupported" to true, // Android always supports notifications
            "isDynamicIslandSupported" to false // Android doesn't have Dynamic Island
        )
        
        // Events
        Events(
            "onActivityUpdate",
            "onUserAction", 
            "onActivityEnd",
            "onError",
            "onAudioRecordingUpdate"
        )
        
        // Event Listener Management
        AsyncFunction("addListener") { eventName: String, promise: Promise ->
            // This is a no-op on Android since events are sent automatically
            // The listener management is handled by the React Native event system
            promise.resolve(mapOf("remove" to { /* No-op for Android */ }))
        }
        
        // MARK: - Live Activity Management Functions
        
        AsyncFunction("startActivity") { config: Map<String, Any>, promise: Promise ->
            coroutineScope.launch {
                try {
                    val activityConfig = parseActivityConfig(config)
                    val result = startActivityUseCase.execute(activityConfig)
                    
                    when (result) {
                        is Result.Success -> {
                            promise.resolve(serializeActivity(result.data))
                        }
                        is Result.Failure -> {
                            promise.reject("ACTIVITY_ERROR", result.error.message, result.error)
                        }
                    }
                } catch (e: Exception) {
                    promise.reject("INVALID_CONFIGURATION", e.message, e)
                }
            }
        }
        
        AsyncFunction("updateActivity") { activityId: String, content: Map<String, Any>, promise: Promise ->
            coroutineScope.launch {
                try {
                    val activityContent = parseActivityContent(content)
                    val updateRequest = ActivityUpdateRequest(activityId, activityContent)
                    val result = updateActivityUseCase.execute(updateRequest)
                    
                    when (result) {
                        is Result.Success -> promise.resolve(true)
                        is Result.Failure -> promise.reject("ACTIVITY_ERROR", result.error.message, result.error)
                    }
                } catch (e: Exception) {
                    promise.reject("INVALID_CONFIGURATION", e.message, e)
                }
            }
        }
        
        AsyncFunction("endActivity") { activityId: String, options: Map<String, Any>?, promise: Promise ->
            coroutineScope.launch {
                try {
                    val finalContent = options?.get("finalContent")?.let { 
                        parseActivityContent(it as? Map<String, Any> ?: emptyMap())
                    }
                    val dismissalPolicyStr = options?.get("dismissalPolicy") as? String ?: "default"
                    val dismissalPolicy = DismissalPolicy.fromString(dismissalPolicyStr)
                    
                    val endRequest = ActivityEndRequest(activityId, finalContent, dismissalPolicy)
                    val result = endActivityUseCase.execute(endRequest)
                    
                    when (result) {
                        is Result.Success -> promise.resolve(true)
                        is Result.Failure -> promise.reject("ACTIVITY_ERROR", result.error.message, result.error)
                    }
                } catch (e: Exception) {
                    promise.reject("INVALID_CONFIGURATION", e.message, e)
                }
            }
        }
        
        AsyncFunction("getActiveActivities") { promise: Promise ->
            coroutineScope.launch {
                val result = liveActivityService.getActiveActivities()
                when (result) {
                    is Result.Success -> {
                        promise.resolve(result.data.map { serializeActivity(it) })
                    }
                    is Result.Failure -> promise.resolve(emptyList<Map<String, Any>>())
                }
            }
        }
        
        AsyncFunction("getActivity") { activityId: String, promise: Promise ->
            coroutineScope.launch {
                val result = liveActivityService.getActivity(activityId)
                when (result) {
                    is Result.Success -> {
                        promise.resolve(result.data?.let { serializeActivity(it) })
                    }
                    is Result.Failure -> promise.resolve(null)
                }
            }
        }
        
        // Utility Functions
        Function("validateActivityConfig") { config: Map<String, Any> ->
            try {
                val activityConfig = parseActivityConfig(config)
                val validationResult = validator.validate(activityConfig)
                
                mapOf(
                    "isValid" to validationResult.isValid,
                    "errors" to validationResult.errors.map { error ->
                        mapOf("field" to error.field, "message" to error.message)
                    }
                )
            } catch (e: Exception) {
                mapOf(
                    "isValid" to false,
                    "errors" to listOf(mapOf("field" to "config", "message" to (e.message ?: "Unknown error")))
                )
            }
        }
        
        OnCreate {
            subscribeToEvents()
        }
        
        // View integration
        View(ExpoLiveActivityView::class) {
            Events("onLoad", "onActivityAction")
        }
        
        OnDestroy {
            // Cleanup resources
        }
    }
    
    // MARK: - Private Helper Methods
    
    private fun subscribeToEvents() {
        coroutineScope.launch {
            eventPublisher.eventFlow.collect { event ->
                handleActivityEvent(event)
            }
        }
    }
    
    private fun handleActivityEvent(event: ActivityEvent) {
        when (event) {
            is ActivityEvent.Started -> {
                sendEvent("onActivityUpdate", mapOf(
                    "type" to "started",
                    "activity" to serializeActivity(event.activity)
                ))
            }
            is ActivityEvent.Updated -> {
                sendEvent("onActivityUpdate", mapOf(
                    "type" to "updated",
                    "activityId" to event.request.activityId,
                    "content" to serializeActivityContent(event.request.content),
                    "timestamp" to event.request.timestamp.time
                ))
            }
            is ActivityEvent.Ended -> {
                sendEvent("onActivityEnd", mapOf(
                    "activityId" to event.request.activityId,
                    "finalContent" to event.request.finalContent?.let { serializeActivityContent(it) },
                    "dismissalPolicy" to event.request.dismissalPolicy.name.lowercase()
                ))
            }
            is ActivityEvent.UserAction -> {
                sendEvent("onUserAction", mapOf(
                    "activityId" to event.activityId,
                    "actionId" to event.actionId,
                    "timestamp" to event.timestamp.time
                ))
            }
            is ActivityEvent.Error -> {
                sendEvent("onError", mapOf(
                    "message" to event.error.message,
                    "code" to event.error.code
                ))
            }
        }
    }
}

// MARK: - Parsing Extensions

private fun parseActivityConfig(config: Map<String, Any>): LiveActivityConfig {
    val id = config["id"] as? String ?: throw IllegalArgumentException("ID가 필요합니다")
    val title = config["title"] as? String ?: throw IllegalArgumentException("제목이 필요합니다")
    val typeString = config["type"] as? String ?: "custom"
    val type = ActivityType.fromString(typeString)
    val contentMap = config["content"] as? Map<String, Any> ?: emptyMap()
    val content = parseActivityContent(contentMap)
    val actionsArray = config["actions"] as? List<Map<String, Any>> ?: emptyList()
    val actions = actionsArray.map { parseActivityAction(it) }
    val expirationTimestamp = config["expirationDate"] as? Double
    val expirationDate = expirationTimestamp?.let { Date((it * 1000).toLong()) }
    val priorityString = config["priority"] as? String ?: "normal"
    val priority = ActivityPriority.fromString(priorityString)
    
    return LiveActivityConfig(
        id = id,
        type = type,
        title = title,
        content = content,
        actions = actions,
        expirationDate = expirationDate,
        priority = priority
    )
}

private fun parseActivityContent(content: Map<String, Any>): ActivityContent {
    val status = content["status"] as? String
    val progress = content["progress"] as? Double
    val message = content["message"] as? String
    val estimatedTime = content["estimatedTime"] as? Int
    val customData = content["customData"] as? Map<String, Any> ?: emptyMap()
    
    return ActivityContent(
        status = status,
        progress = progress,
        message = message,
        estimatedTime = estimatedTime,
        customData = customData
    )
}

private fun parseActivityAction(action: Map<String, Any>): ActivityAction {
    val id = action["id"] as? String ?: throw IllegalArgumentException("액션 ID가 필요합니다")
    val title = action["title"] as? String ?: throw IllegalArgumentException("액션 제목이 필요합니다")
    val icon = action["icon"] as? String
    val isDestructive = action["destructive"] as? Boolean ?: false
    val deepLink = action["deepLink"] as? String
    
    return ActivityAction(
        id = id,
        title = title,
        icon = icon,
        isDestructive = isDestructive,
        deepLink = deepLink
    )
}

// MARK: - Serialization Extensions

private fun serializeActivity(activity: LiveActivityInstance): Map<String, Any> {
    return mapOf(
        "id" to activity.id,
        "config" to serializeActivityConfig(activity.config),
        "isActive" to activity.isActive,
        "createdAt" to (activity.createdAt.time / 1000.0),
        "updatedAt" to (activity.updatedAt.time / 1000.0),
        "nativeActivityId" to (activity.nativeActivityId ?: "")
    )
}

private fun serializeActivityConfig(config: LiveActivityConfig): Map<String, Any> {
    val result = mutableMapOf<String, Any>(
        "id" to config.id,
        "type" to config.type.name.lowercase(),
        "title" to config.title,
        "content" to serializeActivityContent(config.content),
        "actions" to config.actions.map { serializeActivityAction(it) },
        "priority" to config.priority.name.lowercase()
    )
    config.expirationDate?.let { result["expirationDate"] = it.time / 1000.0 }
    return result
}

private fun serializeActivityContent(content: ActivityContent): Map<String, Any> {
    val result = mutableMapOf<String, Any>()
    content.status?.let { result["status"] = it }
    content.progress?.let { result["progress"] = it }
    content.message?.let { result["message"] = it }
    content.estimatedTime?.let { result["estimatedTime"] = it }
    if (content.customData.isNotEmpty()) result["customData"] = content.customData
    return result
}

private fun serializeActivityAction(action: ActivityAction): Map<String, Any> {
    return mapOf(
        "id" to action.id,
        "title" to action.title,
        "icon" to (action.icon ?: ""),
        "destructive" to action.isDestructive,
        "deepLink" to (action.deepLink ?: "")
    )
}

File: ios/ExpoLiveActivityModule.swift
import ExpoModulesCore
import Foundation

/// Expo Live Activity 모듈 - Presentation Layer
/// Clean Architecture: 외부 인터페이스(JavaScript)와 내부 비즈니스 로직을 연결
public class ExpoLiveActivityModule: Module {
    
    // MARK: - Dependencies (Lazy Initialization)
    
    private lazy var activityKitManager: ActivityKitManagerProtocol = {
        if #available(iOS 16.1, *) {
            return ActivityKitManager()
        } else {
            return MockActivityKitManager()
        }
    }()
    
    private lazy var repository: ActivityRepositoryProtocol = {
        // 실제 앱에서는 UserDefaults 기반 repository 사용
        // 개발/테스트 시에는 InMemory repository 사용 가능
        return UserDefaultsActivityRepository()
    }()
    
    private lazy var eventPublisher: EventPublisherProtocol = {
        return ActivityEventPublisher()
    }()
    
    private lazy var validator: ActivityConfigValidatorProtocol = {
        return ActivityConfigValidator()
    }()
    
    private lazy var liveActivityService: LiveActivityServiceProtocol = {
        return LiveActivityService(
            activityKitManager: activityKitManager,
            repository: repository,
            eventPublisher: eventPublisher
        )
    }()
    
    private lazy var dynamicIslandService: DynamicIslandServiceProtocol = {
        return DynamicIslandService(
            activityKitManager: activityKitManager,
            repository: repository
        )
    }()
    
    private lazy var pushService: PushNotificationServiceProtocol = {
        return PushNotificationService()
    }()
    
    // Use Cases
    private lazy var startActivityUseCase: StartActivityUseCaseProtocol = {
        return StartActivityUseCase(
            activityService: liveActivityService,
            repository: repository,
            validator: validator,
            eventPublisher: eventPublisher
        )
    }()
    
    private lazy var updateActivityUseCase: UpdateActivityUseCaseProtocol = {
        return UpdateActivityUseCase(
            activityService: liveActivityService,
            repository: repository,
            eventPublisher: eventPublisher
        )
    }()
    
    private lazy var endActivityUseCase: EndActivityUseCaseProtocol = {
        return EndActivityUseCase(
            activityService: liveActivityService,
            repository: repository,
            eventPublisher: eventPublisher
        )
    }()
    
    // MARK: - Module Definition
    
    public func definition() -> ModuleDefinition {
        Name("ExpoLiveActivity")
        
        // Constants
        Constants([
            "isSupported": self.activityKitManager.isLiveActivitySupported(),
            "isDynamicIslandSupported": self.dynamicIslandService.isDynamicIslandSupported()
        ])
        
        // Events
        Events([
            "onActivityUpdate",
            "onUserAction", 
            "onActivityEnd",
            "onError",
            "onAudioRecordingUpdate"
        ])
        
        // Event Listener Management
        AsyncFunction("addListener") { (eventName: String) -> [String: Any] in
            // This is a no-op on iOS since events are sent automatically
            // The listener management is handled by the React Native event system
            return ["remove": { /* No-op for iOS */ }]
        }
        
        // MARK: - Live Activity Management Functions
        
        /// Live Activity 시작
        AsyncFunction("startActivity") { [weak self] (config: [String: Any]) -> [String: Any] in
            guard let self = self else {
                throw ActivityError.unknown("모듈이 해제됨")
            }
            
            do {
                let activityConfig = try self.parseActivityConfig(from: config)
                let result = await self.startActivityUseCase.execute(activityConfig)
                
                switch result {
                case .success(let activity):
                    return self.serializeActivity(activity)
                case .failure(let error):
                    throw error
                }
            } catch {
                throw ActivityError.invalidConfiguration(error.localizedDescription)
            }
        }
        
        /// Live Activity 업데이트
        AsyncFunction("updateActivity") { [weak self] (activityId: String, content: [String: Any]) -> Bool in
            guard let self = self else {
                throw ActivityError.unknown("모듈이 해제됨")
            }
            
            do {
                let activityContent = try self.parseActivityContent(from: content)
                let updateRequest = ActivityUpdateRequest(activityId: activityId, content: activityContent)
                let result = await self.updateActivityUseCase.execute(updateRequest)
                
                switch result {
                case .success:
                    return true
                case .failure(let error):
                    throw error
                }
            } catch {
                throw ActivityError.invalidConfiguration(error.localizedDescription)
            }
        }
        
        /// Live Activity 종료
        AsyncFunction("endActivity") { [weak self] (activityId: String, options: [String: Any]?) -> Bool in
            guard let self = self else {
                throw ActivityError.unknown("모듈이 해제됨")
            }
            
            let finalContent: ActivityContent?
            let dismissalPolicy: DismissalPolicy
            
            if let options = options {
                finalContent = try? self.parseActivityContent(from: options["finalContent"] as? [String: Any] ?? [:])
                dismissalPolicy = DismissalPolicy(rawValue: options["dismissalPolicy"] as? String ?? "default") ?? .default
            } else {
                finalContent = nil
                dismissalPolicy = .default
            }
            
            let endRequest = ActivityEndRequest(
                activityId: activityId,
                finalContent: finalContent,
                dismissalPolicy: dismissalPolicy
            )
            
            let result = await self.endActivityUseCase.execute(endRequest)
            switch result {
            case .success:
                return true
            case .failure(let error):
                throw error
            }
        }
        
        /// 활성 Activity 목록 조회
        AsyncFunction("getActiveActivities") { [weak self] () -> [[String: Any]] in
            guard let self = self else { return [] }
            
            let result = await self.liveActivityService.getActiveActivities()
            switch result {
            case .success(let activities):
                return activities.map(self.serializeActivity)
            case .failure:
                return []
            }
        }
        
        /// 특정 Activity 조회
        AsyncFunction("getActivity") { [weak self] (activityId: String) -> [String: Any]? in
            guard let self = self else { return nil }
            
            let result = await self.liveActivityService.getActivity(id: activityId)
            switch result {
            case .success(let activity):
                return activity.map(self.serializeActivity)
            case .failure:
                return nil
            }
        }
        
        // MARK: - Dynamic Island Functions
        
        /// Dynamic Island 업데이트
        AsyncFunction("updateDynamicIsland") { [weak self] (activityId: String, content: [String: Any]) -> Bool in
            guard let self = self else {
                throw ActivityError.unknown("모듈이 해제됨")
            }
            
            do {
                let dynamicIslandContent = try self.parseDynamicIslandContent(from: content)
                let updateRequest = DynamicIslandUpdateRequest(
                    activityId: activityId,
                    content: dynamicIslandContent
                )
                
                let result = await self.dynamicIslandService.updateDynamicIsland(updateRequest)
                switch result {
                case .success:
                    return true
                case .failure(let error):
                    throw error
                }
            } catch {
                throw ActivityError.invalidConfiguration(error.localizedDescription)
            }
        }
        
        // MARK: - Push Notification Functions
        
        /// Push 토큰 등록
        AsyncFunction("registerPushToken") { [weak self] (token: String) -> Bool in
            guard let self = self else {
                throw ActivityError.unknown("모듈이 해제됨")
            }
            
            let result = await self.pushService.registerPushToken(token)
            switch result {
            case .success:
                return true
            case .failure(let error):
                throw error
            }
        }
        
        /// 원격 업데이트 요청
        AsyncFunction("requestRemoteUpdate") { [weak self] (activityId: String, content: [String: Any], pushToken: String) -> Bool in
            guard let self = self else {
                throw ActivityError.unknown("모듈이 해제됨")
            }
            
            do {
                let activityContent = try self.parseActivityContent(from: content)
                let result = await self.pushService.requestRemoteUpdate(
                    activityId: activityId,
                    content: activityContent,
                    pushToken: pushToken
                )
                
                switch result {
                case .success:
                    return true
                case .failure(let error):
                    throw error
                }
            } catch {
                throw ActivityError.invalidConfiguration(error.localizedDescription)
            }
        }
        
        // MARK: - Utility Functions
        
        /// Push 알림 권한 상태 확인
        AsyncFunction("getPushAuthorizationStatus") { [weak self] () -> String in
            guard let self = self else { return "notDetermined" }
            
            let status = await self.pushService.getPushAuthorizationStatus()
            return status.rawValue
        }
        
        /// Activity 설정 검증
        Function("validateActivityConfig") { [weak self] (config: [String: Any]) -> [String: Any] in
            guard let self = self else {
                return ["isValid": false, "errors": [["field": "module", "message": "모듈이 해제됨"]]]
            }
            
            do {
                let activityConfig = try self.parseActivityConfig(from: config)
                let validationResult = self.validator.validate(activityConfig)
                
                return [
                    "isValid": validationResult.isValid,
                    "errors": validationResult.errors.map { error in
                        ["field": error.field, "message": error.message]
                    }
                ]
            } catch {
                return [
                    "isValid": false,
                    "errors": [["field": "config", "message": error.localizedDescription]]
                ]
            }
        }
        
        // MARK: - Module Lifecycle
        
        OnCreate {
            // 이벤트 스트림 구독
            self.subscribeToEvents()
        }
        
        OnDestroy {
            // 리소스 정리
            // EventPublisher는 deinit에서 자동으로 정리됨
        }
    }
    
    // MARK: - Private Helper Methods
    
    private func subscribeToEvents() {
        Task {
            let eventStream = liveActivityService.eventStream()
            for await event in eventStream {
                handleActivityEvent(event)
            }
        }
    }
    
    private func handleActivityEvent(_ event: ActivityEvent) {
        switch event {
        case .started(let activity):
            sendEvent("onActivityUpdate", [
                "type": "started",
                "activity": serializeActivity(activity)
            ])
            
        case .updated(let request):
            sendEvent("onActivityUpdate", [
                "type": "updated",
                "activityId": request.activityId,
                "content": serializeActivityContent(request.content),
                "timestamp": request.timestamp.timeIntervalSince1970
            ])
            
        case .ended(let request):
            sendEvent("onActivityEnd", [
                "activityId": request.activityId,
                "finalContent": request.finalContent.map(serializeActivityContent),
                "dismissalPolicy": request.dismissalPolicy.rawValue
            ])
            
        case .userAction(let action):
            sendEvent("onUserAction", [
                "activityId": action.activityId,
                "actionId": action.actionId,
                "timestamp": action.timestamp.timeIntervalSince1970
            ])
            
        case .error(let error):
            sendEvent("onError", [
                "message": error.localizedDescription,
                "code": error.errorCode
            ])
        }
    }
}

// MARK: - Parsing Extensions

extension ExpoLiveActivityModule {
    
    private func parseActivityConfig(from dict: [String: Any]) throws -> LiveActivityConfig {
        guard let id = dict["id"] as? String else {
            throw ActivityError.invalidConfiguration("ID가 필요합니다")
        }
        
        guard let title = dict["title"] as? String else {
            throw ActivityError.invalidConfiguration("제목이 필요합니다")
        }
        
        let typeString = dict["type"] as? String ?? "custom"
        let type = ActivityType(rawValue: typeString) ?? .custom
        
        let contentDict = dict["content"] as? [String: Any] ?? [:]
        let content = try parseActivityContent(from: contentDict)
        
        let actionsArray = dict["actions"] as? [[String: Any]] ?? []
        let actions = try actionsArray.map(parseActivityAction)
        
        let expirationDate: Date?
        if let timestamp = dict["expirationDate"] as? Double {
            expirationDate = Date(timeIntervalSince1970: timestamp)
        } else {
            expirationDate = nil
        }
        
        let priorityString = dict["priority"] as? String ?? "normal"
        let priority = ActivityPriority(rawValue: priorityString) ?? .normal
        
        return LiveActivityConfig(
            id: id,
            type: type,
            title: title,
            content: content,
            actions: actions,
            expirationDate: expirationDate,
            priority: priority
        )
    }
    
    private func parseActivityContent(from dict: [String: Any]) throws -> ActivityContent {
        let status = dict["status"] as? String
        let progress = dict["progress"] as? Double
        let message = dict["message"] as? String
        let estimatedTime = dict["estimatedTime"] as? Int
        let customData = dict["customData"] as? [String: Any] ?? [:]
        
        return ActivityContent(
            status: status,
            progress: progress,
            message: message,
            estimatedTime: estimatedTime,
            customData: customData
        )
    }
    
    private func parseActivityAction(from dict: [String: Any]) throws -> ActivityAction {
        guard let id = dict["id"] as? String else {
            throw ActivityError.invalidConfiguration("액션 ID가 필요합니다")
        }
        
        guard let title = dict["title"] as? String else {
            throw ActivityError.invalidConfiguration("액션 제목이 필요합니다")
        }
        
        let icon = dict["icon"] as? String
        let isDestructive = dict["destructive"] as? Bool ?? false
        let deepLink = dict["deepLink"] as? String
        
        return ActivityAction(
            id: id,
            title: title,
            icon: icon,
            isDestructive: isDestructive,
            deepLink: deepLink
        )
    }
    
    private func parseDynamicIslandContent(from dict: [String: Any]) throws -> DynamicIslandContent {
        let compactLeading = (dict["compactLeading"] as? [String: Any]).flatMap(parseDynamicIslandElement)
        let compactTrailing = (dict["compactTrailing"] as? [String: Any]).flatMap(parseDynamicIslandElement)
        let minimal = (dict["minimal"] as? [String: Any]).flatMap(parseDynamicIslandElement)
        
        // expanded 콘텐츠는 현재 구현에서는 제외 (복잡도로 인해)
        
        return DynamicIslandContent(
            compactLeading: compactLeading,
            compactTrailing: compactTrailing,
            minimal: minimal,
            expanded: nil
        )
    }
    
    private func parseDynamicIslandElement(from dict: [String: Any]) -> DynamicIslandElement? {
        guard let content = dict["content"] as? String else { return nil }
        
        let typeString = dict["type"] as? String ?? "text"
        let type: DynamicIslandElement.ElementType
        
        switch typeString {
        case "text": type = .text
        case "icon": type = .icon
        case "emoji": type = .emoji
        case "progress": type = .progress
        default: type = .text
        }
        
        let color = dict["color"] as? String
        
        return DynamicIslandElement(
            type: type,
            content: content,
            color: color,
            font: nil
        )
    }
}

// MARK: - Serialization Extensions

extension ExpoLiveActivityModule {
    
    private func serializeActivity(_ activity: LiveActivityInstance) -> [String: Any] {
        return [
            "id": activity.id,
            "config": serializeActivityConfig(activity.config),
            "isActive": activity.isActive,
            "createdAt": activity.createdAt.timeIntervalSince1970,
            "updatedAt": activity.updatedAt.timeIntervalSince1970,
            "nativeActivityId": activity.nativeActivityId as Any
        ]
    }
    
    private func serializeActivityConfig(_ config: LiveActivityConfig) -> [String: Any] {
        return [
            "id": config.id,
            "type": config.type.rawValue,
            "title": config.title,
            "content": serializeActivityContent(config.content),
            "actions": config.actions.map(serializeActivityAction),
            "expirationDate": config.expirationDate?.timeIntervalSince1970 as Any,
            "priority": config.priority.rawValue
        ]
    }
    
    private func serializeActivityContent(_ content: ActivityContent) -> [String: Any] {
        return [
            "status": content.status as Any,
            "progress": content.progress as Any,
            "message": content.message as Any,
            "estimatedTime": content.estimatedTime as Any,
            "customData": content.customData
        ]
    }
    
    private func serializeActivityAction(_ action: ActivityAction) -> [String: Any] {
        return [
            "id": action.id,
            "title": action.title,
            "icon": action.icon as Any,
            "destructive": action.isDestructive,
            "deepLink": action.deepLink as Any
        ]
    }
}

// MARK: - Error Extension

extension ActivityError {
    var errorCode: String {
        switch self {
        case .invalidConfiguration:
            return "INVALID_CONFIGURATION"
        case .activityNotFound:
            return "ACTIVITY_NOT_FOUND"
        case .alreadyStarted:
            return "ALREADY_STARTED"
        case .systemNotSupported:
            return "SYSTEM_NOT_SUPPORTED"
        case .permissionDenied:
            return "PERMISSION_DENIED"
        case .networkError:
            return "NETWORK_ERROR"
        case .unknown:
            return "UNKNOWN_ERROR"
        }
    }
}

extension PushAuthorizationStatus {
    var rawValue: String {
        switch self {
        case .notDetermined: return "notDetermined"
        case .denied: return "denied"
        case .authorized: return "authorized"
        case .provisional: return "provisional"
        case .ephemeral: return "ephemeral"
        }
    }
}

// MARK: - Mock Implementation for iOS < 16.1

/// iOS 16.1 미만에서 사용할 Mock ActivityKit Manager
private class MockActivityKitManager: ActivityKitManagerProtocol {
    func isLiveActivitySupported() -> Bool { false }
    func startNativeActivity(_ config: LiveActivityConfig) async -> Result<String, ActivityError> {
        return .failure(.systemNotSupported)
    }
    func updateNativeActivity(nativeId: String, content: ActivityContent) async -> Result<Void, ActivityError> {
        return .failure(.systemNotSupported)
    }
    func endNativeActivity(nativeId: String, finalContent: ActivityContent?, dismissalPolicy: DismissalPolicy) async -> Result<Void, ActivityError> {
        return .failure(.systemNotSupported)
    }
    func nativeEventStream() -> AsyncStream<ActivityEvent> {
        return AsyncStream { _ in }
    }
}